### PR TITLE
Add GET /restaurants endpoint

### LINKED ISSUE #9: Fetch restaurants from the backend
1. Add `GET /restaurants` endpoint that responds with a hardcoded json.

```json
[
  {
    "id": "1",
    "name": "Restaurant 1",
    "address": "Address 1",
    "logo": "https://axwwgrkdco.cloudimg.io/v7/__gmpics__/f23bb032d19c444cb062d0016d2525c2",
    "rating": 4
  },
  {
    "id": "2",
    "name": "Restaurant 2",
    "address": "Address 2",
    "logo": "https://axwwgrkdco.cloudimg.io/v7/__gmpics__/502c3efff32b4beba04c7b2b7bdc324c",
    "rating": 5
  }
]
```

2. Fetch restaurants on the frontend form the locally run backend

### PR DESCRIPTION
**Pull request by adam-lzk**
I made several versions, I don't know how better:
• GetRestaurants1.js accepts GET requests, but listens to all endpoints, not just /restaurants
• GetRestaurants2.js accepts GET requests, listens to /restaurants endpoint, but it returns an ugly formatted json
• GetRestaurants3.js accepts GET requests, listens to /restaurants endpoint, reads and returns json from an external file

### CODE REVIEW

File: src/pages/Restaurants.jsx [hunk 1]
```diff
@@ -142,6 +30,19 @@ const CardDescription = ({ description, rating, priceRange, cuisine }) => {
 }
 
 export const RestaurantsPage = () => {
+  const [cardsData, setCardsData] = useState()
```
Comment (dmlaziuk) [inline src/pages/Restaurants.jsx]: Please use an empty array as an initial state.
```suggestion
  const [cardsData, setCardsData] = useState([])
```

File: src/endpoints/GetRestaurants.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  res.setHeader("Access-Control-Max-Age", 2592000)  // 30 days
+  res.setHeader('Access-Control-Allow-Headers', '*')
+
+  if (req.method === "OPTIONS") {
+    res.writeHead(200)
+    res.end()
+    return
+  }
+
+  const cardsData = [
```
Comment (dmlaziuk) [inline src/endpoints/GetRestaurants.js]: Please move this data to a separate file called `data.json` and import it here.

File: src/endpoints/GetRestaurants.js
Comment (dmlaziuk) [inline src/endpoints/GetRestaurants.js:1]: Please name folder `backend`, place it in the root of the project and rename the file to `main.js`.

File: package.json [hunk 3]
```diff
@@ -13,6 +13,7 @@
     "@emotion/react": "^11.11.1",
     "@emotion/styled": "^11.11.0",
     "antd": "^5.9.0",
+    "express": "^4.18.2",
```
Comment (dmlaziuk) [inline package.json]: We're not using express in this PR. Please remove unused dependency.

File: src/endpoints/GetRestaurants.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      "id":11,
+      "rating":4.7,
+      "priceRange":3,
+      "cuisine":"Creative",
+      "alt":"Fiorentina",
+      "src":"https://axwwgrkdco.cloudimg.io/v7/__gmpics__/335bb0d55d184d40965b042dac8817ea"
+    }
+  ]
+
+  if (req.url === "/restaurants") {
```
Comment (dmlaziuk) [inline src/endpoints/GetRestaurants.js]: We should compare only pathname, not the whole `req.url`.

Please add a console.log and make sure you understand what [`URL`](https://en.wikipedia.org/wiki/URL) is. What are the parts of the URL?

```js
console.log({ url: req.url, newURL: new URL(req.url, `http://${req.headers.host}`) })
```
Screenshot 2023-11-12 at 00 37 14

```sh
curl "http://localhost:8080/restaurants?page=1&per_page=10"
```

**Review by dmlaziuk — CHANGES_REQUESTED** [review]

File: src/pages/Restaurants.jsx (hunk 1 above)
Comment (adam-lzk) [reply src/pages/Restaurants.jsx]: Fixed

File: src/endpoints/GetRestaurants.js (hunk 2 above)
Comment (adam-lzk) [reply src/endpoints/GetRestaurants.js]: Fixed

File: src/endpoints/GetRestaurants.js
Comment (adam-lzk) [reply src/endpoints/GetRestaurants.js:1]: Fixed

File: package.json (hunk 3 above)
Comment (adam-lzk) [reply package.json]: Fixed

File: src/endpoints/GetRestaurants.js (hunk 4 above)
Comment (adam-lzk) [reply src/endpoints/GetRestaurants.js]: Fixed

File: package-lock.json
Comment (dmlaziuk) [inline package-lock.json:1]: Please uncommit this file from the PR

File: backend/main.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  //console.log({ pathname: newURL.pathname })
+  //console.log({ 'newURL': newURL })
+  // to check - curl "http://localhost:8080/restaurants?page=1&per_page=10"
+
+  //console.log({ 'searchParams': newURL.searchParams })
+  //console.log({ 'searchParams.get(key)': newURL.searchParams.get('id') })
+
+  if (newURL.pathname === "/restaurants") {
+    if (newURL.searchParams.has('id')) {
+      const requiredCard = cardsData.filter(card => card.id.toString() === newURL.searchParams.get('id'))
```
Comment (dmlaziuk) [inline backend/main.js]: `filter` returns a filtered array, while `find` returns an array element.
```suggestion
    const id = newURL.searchParams.get('id')
    if (id) {
      const requiredCard = cardsData.find(card => card.id.toString() === id)
```

File: backend/main.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+      if (requiredCard.length === 1) {
```
Comment (dmlaziuk) [inline backend/main.js]:
```suggestion
      if (requiredCard) {
```

File: backend/main.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        //console.log({ requiredCardID: requiredCard[0].id })
+        res.setHeader("Content-Type", "application/json")
+        res.writeHead(200)
+        res.end(JSON.stringify(requiredCard))
```
Comment (dmlaziuk) [inline backend/main.js]: in the future we'll return just one element, not an array with a single element
```suggestion
        res.end(JSON.stringify([requiredCard]))
```

File: src/pages/Restaurants.jsx [hunk 8]
```diff
@@ -142,6 +30,23 @@ const CardDescription = ({ description, rating, priceRange, cuisine }) => {
 }
 
 export const RestaurantsPage = () => {
+  const [cardsData, setCardsData] = useState([])
+
+  const websitePort = window.location.port
+  const serverPort = '8080'
+  const url = window.location.href.replace(websitePort, serverPort)
```
Comment (dmlaziuk) [inline src/pages/Restaurants.jsx]: This not what we want.
Api endpoint is something that we always know. For example `GET /restaurants` and we can call this endpoint from any page or even any place in the codebase.

```suggestion
  const { restaurants } = useRestaurants()
```

Terefore please create a new hook that returns restaurants in src/hooks/useRestaurants.js:
```js
export const useRestaurants = (params) => {
  ...all the fetch logic here

  return { restaurants }
}
```

**Review by dmlaziuk — CHANGES_REQUESTED** [review]

File: package-lock.json
Comment (adam-lzk) [reply package-lock.json:1]: Fixed

File: backend/main.js (hunk 5 above)
Comment (adam-lzk) [reply backend/main.js]: Fixed

File: backend/main.js (hunk 6 above)
Comment (adam-lzk) [reply backend/main.js]: Fixed

Comment (adam-lzk) [reply backend/main.js]: Fixed

File: backend/main.js (hunk 7 above)
Comment (adam-lzk) [reply backend/main.js]: Fixed

File: src/pages/Restaurants.jsx (hunk 8 above)
Comment (adam-lzk) [reply src/pages/Restaurants.jsx]: Fixed

File: package-lock.json
Comment (dmlaziuk) [inline package-lock.json:1]: Don't delete this file from the repository.
I requested just to remove unnecessary changes.

You can always pick this file from any time in the git history by just `git checkout some_commit_hash filename`

**Review by dmlaziuk — CHANGES_REQUESTED** [review]

**Review by dmlaziuk — APPROVED** [review]
